### PR TITLE
Customise Homepage

### LINKED ISSUE #75: Customisable homepage
The homepage should be editable to be able to choose which articles/surveys/polls/quizzes appear on the homepage.

Currently this is done through editing the article by going to Promote > FEATURING IN HOMEPAGE

Instead, the desired functionality is for this to be done by editing the homepage and not by having to go to each individual article. Selecting items for the homepage should be done in a similar way as "Recommended Articles" are chosen where you can add articles/surveys/polls/quizzes through a search and be in a specific order.

### PR DESCRIPTION
**Pull request by sheralim012**
Fixed #75 

* Added featured content support on Homepage in wagtail admin
* Added banners support on Homepage wagtail admin
* Made menu flat (no sub-navigation)
* Set default to `True` for **show in menus** for sections and articles

### CODE REVIEW

**Comment by sheralim012** [comment]
Hey @dannyparsons

Screenshot 2021-05-30 at 8 40 22 PM
menu support included, you can create a new menu from Settings>Main Menu, just make sure the page you are adding is eligible to be shown in the menu by checking the promote tab. Levels for the menu could also be selected from menu settings.

Screenshot 2021-05-30 at 8 40 55 PM
featured content support included, you can add featured contents on homepage.

Screenshot 2021-05-30 at 9 09 36 PM
footer support included, you can create new snippet for footer from wagtail menu.

**Comment by dannyparsons** [comment]
Thanks for this. This is looking good, a few points:

- Can we set **Show in menus** to be checked by default for articles and sections? I can see this being confusing for CMS users when creating the menu and not knowing why an item doesn't appear.
- In Settings > Main Menu > Settings, can you set the default to be **1. No Sub-navigation (flat)** - this is what we want the sites to use.
- The Featured Content works well. When I run it, it still shows all the articles first and then the featured content below. We only want the featured content to show now, not all the articles. Not sure if that's just for me?
- Footers also display. Probably need a design and functionality discussion around these, but not for this PR

**Comment by sheralim012** [comment]
Screenshot 2021-06-01 at 1 12 05 PM
* banners included, with clickable images for internal and external pages
* made **Show in menus** default to checked
* made menu sub navigation to **flat**
* removed old articles from homepage

**Comment by dannyparsons** [comment]
Thanks. Can you swap Banners and Featured Content sections in the CMS when editing the homepage? Banners appear above so it will be more logical for it to be above in the CMS as well.

**Comment by sheralim012** [comment]
unified migration files, so before pulling new code run the following commands
```
make ssh
cd /app/
python manage.py migrate home 0004
```

then you can pull new code and run new migrations

**Comment by dannyparsons** [comment]
Looks good. Minor points:

- Could you make the Banner image mandatory? It isn't in v1 but I think it needs to be otherwise there isn't anything to display
- Could you add the help text on hover for the banner fields? (see below) Something like:
Banner image: Image to display as the banner
Link page: Optional page to which the banner will link to
External link: Optional external link which a banner will link to eg https://www.google.com

image

**Comment by sheralim012** [comment]
@dannyparsons addressed both points.

**Comment by dannyparsons** [comment]
Thanks. 

@istride raised some important points on use of banners https://github.com/unicef/iogt/issues/75#issuecomment-852264394 I think that needs to be talked through with him next.

**Comment by sheralim012** [comment]
@dannyparsons banner and footer are implemented as pages now, kindly check.

**Comment by sheralim012** [comment]
fixed migration files, so before pulling new code run the following commands
```
make ssh
cd /app/
python manage.py migrate home 0004
```
then you can pull new code and run new migrations

**Comment by dannyparsons** [comment]
Thanks.

I was assuming that banners and footers would still be selected in a field from the homepage like the featured content. This would be desirable for the CMS editors. It would mean easier control over which banners are included and in what order. Banners could be created under a Banners Index Page (like in v1), instead of having to be children of the homepage. The same for footer. I think that would make it CMS editing clearer.

**Comment by sheralim012** [comment]
Hey @dannyparsons,

I have implemented Banner and Footer Index Pages (to hold all banners and footers respectively) and Banner and Footer for individual entity just like V1 and I have also added the ability to select banners for the home page just like featured content. One thing I want to confirm is that should selecting footer also be an option on home page?

**Comment by sheralim012** [comment]
fixed migration files, so before pulling new code run the following commands
```
make ssh
cd /app/
python manage.py migrate home 0004
```
then you can pull new code and run new migrations

**Comment by dannyparsons** [comment]
Thanks. I like this implementation because:

- Banners are defined in their own place (Banner Index Page) so don't clutter the site structure. 
- The selection of banners as a field on the homepage means the things that will appear on the homepage are defined in one place. Also Laurie mentioned that the ordering of banners on the Banner Index Page in v1 is quite confusing. Ordering them on the homepage field is a much nice interface.
- Banners don't have to be put back into draft for them to be hidden (they are just removed from the homepage)
- If we ever wanted to use banners on other pages we could easily do this.

Footers are maybe different because they don't just appear on the homepage they appear across the site. So conceptually they don't belong on the homepage. But it would be much better to have the ordering of them done through the page selection way instead of the Footer Index Page.

I think we need input from @istride before deciding.

**Comment by sheralim012** [comment]
@dannyparsons @istride reverted footer changes and currently, this PR contains featured content and banner. 

Also fixed migration files, so before pulling new code run the following commands
```
make ssh
cd /app/
python manage.py migrate home 0004
```
then you can pull new code and run new migrations

File: home/templates/home/tags/banners_list.html [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    {% for banner in banners %}
+    <div class="bannerContainer">
+        <div class="desktopBannerContainer">
+            <div class="image">
+              {% if banner.specific.banner_link_page %}
+                <a href="{% pageurl banner.specific.banner_link_page %}">
+              {% elif banner.specific.external_link %}
+                <a href="{{ banner.specific.external_link }}" target="_blank">
+              {% else %}
+                <a href="#"></a>
```
Comment (istride) [inline home/templates/home/tags/banners_list.html:13]: I think the anchor tag should not be closed here.

**Review by istride — CHANGES_REQUESTED** [review]

File: home/models.py [hunk 2]
```diff
@@ -15,12 +17,41 @@
...
+            InlinePanel('home_page_banners', label=_("Home Page Banner")),
+        ], heading=_('Home Page Banners')),
+        MultiFieldPanel([
+            InlinePanel('featured_content', label=_("Featured Content")),
+        ], heading=_('Featured Content')),
+    ]
+
     def get_context(self, request):
         context = super().get_context(request)
         context['articles'] = self.get_descendants().type(Article)
```
[the remark this answers is not in the document]
Comment (sheralim012) [reply home/models.py]: done

File: home/models.py [hunk 3]
```diff
@@ -15,12 +17,41 @@
...
+        context['banners'] = [home_page_banner.banner_page for home_page_banner in self.home_page_banners.all()]
```
[the remark this answers is not in the document]
Comment (sheralim012) [reply home/models.py]: fetching only live content now

File: home/templates/home/tags/banners_list.html (hunk 1 above)
Comment (sheralim012) [reply home/templates/home/tags/banners_list.html]: removed unnecessary closing tag

**Comment by sheralim012** [comment]
@istride changes addressed.

**Review by istride — APPROVED** [review]

**Comment by istride** [comment]
Merged manually as I had to merge changes from develop and recreate the database migrations.

**Pull request closed without merging**